collector/btrfs_linux.go

```diff
@@ -52,16 +56,214 @@ func NewBtrfsCollector(logger log.Logger) (Collector, error) {
 func (c *btrfsCollector) Update(ch chan<- prometheus.Metric) error {
 	stats, err := c.fs.Stats()
 	if err != nil {
-		return fmt.Errorf("failed to retrieve Btrfs stats: %w", err)
+		return fmt.Errorf("failed to retrieve Btrfs stats from procfs: %w", err)
+	}
+
+	ioctlStatsMap, err := c.getIoctlStats()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve Btrfs stats with ioctl: %w", err)
 	}
 
 	for _, s := range stats {
-		c.updateBtrfsStats(ch, s)
+		// match up procfs and ioctl info by filesystem UUID
+		ioctlStats := ioctlStatsMap[s.UUID]
+		c.updateBtrfsStats(ch, s, ioctlStats)
 	}
 
 	return nil
 }
 
+type ioctlFsDeviceStats struct {
+	path string
+	uuid string
+
+	bytesUsed  uint64
+	totalBytes uint64
+
+	writeErrors      uint64
+	readErrors       uint64
+	flushErrors      uint64
+	corruptionErrors uint64
+	generationErrors uint64
+}
+
+type ioctlFsStats struct {
+	uuid       string
+	mountPoint string
+	devices    []ioctlFsDeviceStats
+}
+
+func (c *btrfsCollector) getIoctlStats() (map[string]*ioctlFsStats, error) {
+	// Instead of introducing more ioctl calls to scan for all btrfs
+	// filesytems re-use our mount point utils to find known mounts
```

Comment:
Does anyone know if bind mounts will causes problems here? I.e. will we end up scanning filesystems multiple times?

Reply:
Yes, it will. This is exactly the largest issue with the existing script collector (from the -contrib repo) and why I wrote my own based on python-btrfs, which will iterate over /sys/fs/btrfs/<uuid> instead; see the  discussion in https://github.com/knorrie/python-btrfs/pull/26.

Reply:
Thanks! Do you know of any good syscalls list all btrfs filesystems, or a way to look up a btrfs filesystem by UUID rather than mountpoint?
I've had a look but I'm new to linux syscalls, and I can barely find any explanations of what each btrfs ioctl cmd is for.

Otherwise I think I'll need to do as the python code seems to do; make the FS_INFO ioctl call for every filesystem and then deduplicate based on UUID.

Reply:
I'm not aware of any ioctls to find/list filesystems (the current list is [here](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/btrfs.h)). For any other questions you shoud really contact the btrfs list (linux-btrfs@vger.kernel.org), everyone else is there as well. :)

Reply:
I've spent some time improving the logic for dealing with the mount point list, which I think ought to solve the issue! Please let me know what you think :)